src/array.c

```diff
@@ -629,156 +642,408 @@ static size_t limit_overallocation(jl_array_t *a, size_t alen, size_t newlen, si
+        array_try_unshare(a);
+    if (idx < n - last) {
+        jl_array_del_at_beg(a, idx, dec, n);
     }
     else {
-        size_t alen = a->nrows;
-        size_t anb = alen*es;
-        if (inc > (a->maxsize-alen)/2 - (a->maxsize-alen)/20) {
-            size_t newlen = a->maxsize==0 ? inc*2 : a->maxsize*2;
-            while (alen+2*inc > newlen-a->offset)
-                newlen *= 2;
-
-            newlen = limit_overallocation(a, alen, newlen, 2*inc);
-            size_t center = (newlen - (alen + inc))/2;
-            array_resize_buffer(a, newlen, alen, center+inc);
-            a->offset = center;
-            a->data = (char*)a->data - incnb;
-        }
-        else {
-            size_t center = (a->maxsize - (alen + inc))/2;
-            char *newdata = (char*)a->data - es*a->offset + es*center;
-            memmove(&newdata[incnb], a->data, anb);
-            a->data = newdata;
-            a->offset = center;
-        }
+        jl_array_del_at_end(a, idx, dec, n);
     }
-    if (a->flags.ptrarray)
-        memset((char*)a->data, 0, incnb);
-#ifdef STORE_ARRAY_LEN
-    a->length += inc;
-#endif
-    a->nrows += inc;
 }
 
 JL_DLLEXPORT void jl_array_del_beg(jl_array_t *a, size_t dec)
 {
-    if (dec == 0) return;
-    if (dec > a->nrows)
+    size_t n = jl_array_nrows(a);
+    if (__unlikely(dec > n))
         jl_bounds_error_int((jl_value_t*)a, dec);
-    if (a->flags.isshared) array_try_unshare(a);
-    size_t es = a->elsize;
-    size_t nb = dec*es;
-    memset(a->data, 0, nb);
-    size_t offset = a->offset;
-    offset += dec;
-    a->data = (char*)a->data + nb;
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_beg(a, 0, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_del_end(jl_array_t *a, size_t dec)
+{
+    size_t n = jl_array_nrows(a);
+    if (__unlikely(n < dec))
+        jl_bounds_error_int((jl_value_t*)a, 0);
+    if (__unlikely(a->flags.isshared))
+        array_try_unshare(a);
+    jl_array_del_at_end(a, n - dec, dec, n);
+}
+
+JL_DLLEXPORT void jl_array_sizehint(jl_array_t *a, size_t sz)
+{
+    size_t n = jl_array_nrows(a);
+    if (sz <= n)
+        return;
+    size_t inc = sz - n;
+    jl_array_grow_end(a, inc);
+    a->nrows = n;
 #ifdef STORE_ARRAY_LEN
-    a->length -= dec;
+    a->length = n;
 #endif
-    a->nrows -= dec;
+}
 
-    // make sure offset doesn't grow forever due to deleting at beginning
-    // and growing at end
-    size_t newoffs = offset;
-    if (offset >= 13*a->maxsize/20) {
-        newoffs = 17*(a->maxsize - a->nrows)/100;
+JL_DLLEXPORT jl_array_t *jl_array_copy(jl_array_t *ary)
+{
+    size_t elsz = ary->elsize;
+    jl_array_t *new_ary = _new_array_(jl_typeof(ary), jl_array_ndims(ary),
+                                      &ary->nrows, !ary->flags.ptrarray, elsz);
+    memcpy(new_ary->data, ary->data, jl_array_len(ary) * elsz);
+    return new_ary;
+}
+
+// Copy element by element until we hit a young object, at which point
+// we can
```

Comment:
continue using `memmove`.

Apparently got interrupted while writing this comment....